### PR TITLE
Torrent file

### PR DESCRIPTION
**Pull request by avrham408**

### CODE REVIEW

File: tests/jupyter/tester.ipynb [hunk 1]
```diff
@@ -0,0 +1,247 @@
+{
```
Comment (adarmaor) [inline tests/jupyter/tester.ipynb]: gitignore

File: torrent/networking.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    try:
+        interested_res = await reader.read(5)
+    except ConnectionResetError:
+        logger.debug("connection with peer end")
+        return False
+    _, choke_status = get_message_type(interested_res)
+    return choke_status 
+
+
+if __name__ == "__main__":
```
Comment (avrham408) [inline torrent/networking.py]: remove this

File: torrent/peer.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+def _valid_peer(ip, port):
+    try:
+        socket.inet_aton(ip)
+    except socket.error:
+        return False
+    if not 1 <= port <= 65535:
+        return False
+    return True
+
```
Comment (avrham408) [inline torrent/peer.py]: too many newlines

File: torrent/pieces.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            elif piece.status == PieceStatus.done:
+                done += 1
+            else:
+                raise Exception(f"{piece.index} have not valid status - {piece.status}")
+                logger.error(exc_info=True)
+        return {PieceStatus.free: free, PieceStatus.done: done, PieceStatus.in_progress: in_progress}
+    
+    async def get_piece(self):
+        while True:
+            if self.lock == False:
```
Comment (avrham408) [inline torrent/pieces.py]: is false

File: torrent/test_files [hunk 5]
```diff
@@ -0,0 +1 @@
+../tests/test_files/
```
Comment (avrham408) [inline torrent/test_files]: remove this file

File: torrent/torrent_file.py [hunk 6]
```diff
@@ -1,64 +1,83 @@
...
     the object contain 2 types of trackers udp or http trackers
     url(domain):str, path:str, type:'http' or 'udp, port:'int'
     """
+    def __init__(self, tracker_type, url, path,  port=80):
+        self.schema = tracker_type
+        self.url = url
+        self.path = path
+        self.port = port
+
+        if tracker_type == 'https':
```
Comment (avrham408) [inline torrent/torrent_file.py]: `self.tracker_type = tracker_type.replace('https', 'http')` (if that doesn't work use re.sub, or use an enum)

File: torrent/torrent_file.py [hunk 7]
```diff
@@ -77,14 +96,14 @@ def read_file(path):
...
-def decode_file(raw_data):
+def decode_raw_data(raw_data):
     """
         the function get bits object and return 
         the file decode in bencoding format
     """
     try: 
         return Decoder(raw_data).decode()
-    except RuntimeError as e:
+    except (RuntimeError, TypeError) as e:
```
Comment (avrham408) [inline torrent/torrent_file.py]: remove as e

File: torrent/torrent_file.py [hunk 8]
```diff
@@ -122,38 +143,73 @@ def multi_file_parse(info_od):
...
-    info_data['pieces'] = od_get_key(info_od, b'pieces', True)
+    pieces = od_get_key(info_od, b'pieces', True)
+    _valid_pieces_is_devide_20(pieces)
+    info_data['pieces'] = pieces
     return info_data 
 
-def create_tracker(bytes_url):
-    pass
+
+def _valid_pieces_is_devide_20(pieces):
```
Comment (avrham408) [inline torrent/torrent_file.py]: is_divided_by_20

File: torrent/torrent_file.py [hunk 9]
```diff
@@ -122,38 +143,73 @@ def multi_file_parse(info_od):
...
+    return True
+
+
+def create_tracker(url):
+    protocol, netloc , path, _, _ = urlsplit(url)
+    protocol 
+    if protocol == 'wss': #temp for deletion in future
+    #TODO support wss protocol
+        return None
+    if protocol != 'udp' and protocol != 'http' and protocol !=  'https':
```
Comment (avrham408) [inline torrent/torrent_file.py]: if protocol not in ['udp', 'http', 'https']

File: torrent/torrent_file.py [hunk 10]
```diff
@@ -122,38 +143,73 @@ def multi_file_parse(info_od):
...
+        logger.warning(f'the trackers {url} not conatin protocol')
+        return None
+    if netloc.find(':') == -1:
```
Comment (avrham408) [inline torrent/torrent_file.py]: if ':' not in netloc

File: torrent/torrent_file.py [hunk 11]
```diff
@@ -122,38 +143,73 @@ def multi_file_parse(info_od):
...
     announce_list = od_get_key(od_torrent, b'announce-list')
-    if not announce_list:
-        return [create_tracker(od_get_key[b'announce'])]
-    else:
+    url_list =  od_get_key(od_torrent, b'url-list')
+    single_announce =  od_get_key(od_torrent, b'announce')
+    trackers = []
+    
+    if announce_list:
+        trackers += [create_tracker(url.decode('utf-8'))  for list_of_url in announce_list for url in list_of_url]
```
Comment (avrham408) [inline torrent/torrent_file.py]: not really understood

File: torrent/networking.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+        6 - request
+        7 - piece
+        8 - cancel
+    """ 
+    #empty
+    if peer_response == b'':
+        logger.debug('we got empty response')
+    #keep alive
+    elif len(peer_response) == 4 and peer_response == b'0000':
+        return -1, 0
```
Comment (adarmaor) [inline torrent/networking.py:62]: what is this?

File: torrent/networking.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+    if not await handshake(reader, writer, torrent_file):
+        logger.debug("connection failed")
+        return False
+    logger.debug(f"{peer} -  handshake sent")
+    if not await bitfield(reader, peer):
+        logger.debug(f"{peer} - bitfield failed")
+    logger.debug(f"{peer} - bitfiled success")
+    message_id = await intersted(reader, writer)
+    if message_id is None or message_id not in (0, 1):
+        logger.debug(f"{peer} - intersted request failed")
```
Comment (adarmaor) [inline torrent/networking.py:88]: don't you want to return in this case? 
in general you can split this function into lots of little functions for readability

File: torrent/networking.py [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+        return False
+    logger.debug(f"{peer} - unchoke")
+    while True:
+        piece = await piece_manager.get_piece()
+        if piece:
+            block_status = await request_block(piece, reader, writer, peer)
+            if not block_status:
+                #TODO sleep and recursion
+                logger.debug("block request failed")
+                piece.lost_piece
```
Comment (adarmaor) [inline torrent/networking.py]: what does this do?

File: torrent/networking.py [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+    block = b''
+    while True:
+        chunk = await reader.read(size)
+        block += chunk 
+        if len(block) >= size:
+            return block[:size]
+        elif chunk == b'':
+            logger.info("chunk of block got empty")
+            return False
+
```
Comment (adarmaor) [inline torrent/networking.py]: too many lines

File: torrent/peer.py [hunk 16]
```diff
@@ -0,0 +1,62 @@
+from queue import Queue
+from time import sleep
+import socket
+import logging
+from enum import Enum
```
Comment (adarmaor) [inline torrent/peer.py:5]: not used

File: torrent/peer.py [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+            sleep(5)
+            self.add_peers(peers)
+
+
+class Peer():
+    def __init__(self, ip, port):
+        self.port = port
+        self.ip = ip
+        self.status = PeerStatus.free
+
```
Comment (adarmaor) [inline torrent/peer.py:46]: too many lines

File: torrent/pieces.py [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+import logging
+import asyncio
+
+logger = logging.getLogger(__name__)
+
+class PieceStatus(Enum):
+    free = 0 
+    done = 2
+    in_progress = 1
+
```
Comment (adarmaor) [inline torrent/pieces.py:14]: 2 lines between objects 
1 line between class methods

File: torrent/pieces.py [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+        logger.debug(f"block add to piece {self.index}")
+        self.blocks.append(block)
+    
+    def __repr__(self):
+        return f"{self.index}:{self.status.name}"
+    
+
+class Piece_Manager():
+    """
+        the class porpuse is to manage the connection from piece to the place in
```
Comment (adarmaor) [inline torrent/pieces.py]: purpose

File: torrent/pieces.py [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+        if type(piece) != Piece:
+            raise Exception(f"add piece to Piece manager failed the object appended in not piece is {self.piece}")
+        self.pieces.append(piece)
+        return True
+
+    def pieces_status(self):
+        in_progress = 0
+        free = 0
+        done = 0
+        for piece in self.pieces:
```
Comment (adarmaor) [inline torrent/pieces.py]: i think its better to create a list of statuses `statuses = map(lambda piece: piece.status, self.pieces)` and then use list.count.
if its important to keep the ones without a valid status use a dict, but i don't think you should be checking that here. you should be checking that wherever you assign a status to a piece

File: torrent/pieces.py [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+    free = 0 
+    done = 2
+    in_progress = 1
+
+class Piece():
+    """The piece object reprsent one piece from pieces in the info of torrent file
+    all piece in size of piece length fro torrent file hoz me except from the last piece
+    the piece come with info hash of 20 bytes in sha1 and if the piece data the came
+    from peer in sha1 not the same like the piece hash the piece is not right"""
+    def __init__(self, piece_hash, length, piece_index):
```
Comment (adarmaor) [inline torrent/pieces.py:20]: i think you should create a function `set_status` that makes sure the status is from PieceStatus enum

File: torrent/torrent_file.py [hunk 22]
```diff
@@ -1,64 +1,83 @@
...
+        self.url = url
+        self.path = path
+        self.port = port
+
+        if tracker_type == 'https':
+            self.tracker_type = 'http'
+        else:
+            self.tracker_type = tracker_type 
+
+    def http_url(self):
```
Comment (adarmaor) [inline torrent/torrent_file.py:51]: this should be a property

File: torrent/torrent_file.py [hunk 23]
```diff
@@ -1,64 +1,83 @@
...
-        if mandat:
-            raise e
-        else:
-            return None
-        
 
 #torrent parse functions
 def valid_torrent_path(path):
     # the function valid torrent object path
+    if type(path) != str:
```
Comment (adarmaor) [inline torrent/torrent_file.py:72]: again, i think you should make the change to python3.8 and add param type validation

File: torrent/torrent_file.py (hunk 8 above)
Comment (adarmaor) [reply torrent/torrent_file.py]: also, validate

File: torrent/tracker.py [hunk 24]
```diff
@@ -0,0 +1,378 @@
...
+import struct 
+import socket
+from enum import Enum
+
+logger = logging.getLogger(__name__)
+
+class UdpError(Enum):
+    unknown_error = 0
+    info_hash = 1
+    not_valid_tracker = 2
```
Comment (adarmaor) [inline torrent/tracker.py]: invalid

File: torrent/tracker.py [hunk 25]
```diff
@@ -0,0 +1,378 @@
...
+    return parsed_peers 
+
+
+def udp_request_connection_response(sock):
+    """
+    errors:
+        0 = restart
+        1 = kill thread 
+        UdpError
+    """
```
Comment (adarmaor) [inline torrent/tracker.py:99]: why are you returning integers here?? what do they mean? add a new enum

File: torrent/tracker.py [hunk 26]
```diff
@@ -0,0 +1,378 @@
...
+        if type(announce_response) == UdpError:
+            return announce_response 
+        interval, peers = announce_response
+        parsed_peers = parse_peers(peers) 
+        peer_manager.add_peers(parsed_peers)
+        logger.info(f"{len(peers)}peers add to peer_manager and go to sleep for {interval} seconds")
+        sleep(interval)
+
+
+def  parse_peers(raw_peers):
```
Comment (adarmaor) [inline torrent/tracker.py]: two spaces between def and function name

File: torrent/tracker.py [hunk 27]
```diff
@@ -0,0 +1,378 @@
...
+        return None
+
+
+def create_connect_req_struct(protocol_id=0X41727101980, transcation_id=randrange(0, 65535), action=0):
+    #the struct for udp tracker request
+    #protocol_id is const, transcation id  is random num , action 0 = connect
+    try:
+        req_struct = struct.pack(">QLL", protocol_id, action, transcation_id)
+    except struct.error:
+        raise Excpetion("the udp req struct creation failed")
```
Comment (adarmaor) [inline torrent/tracker.py]: does this work?

Comment (adarmaor) [reply torrent/tracker.py]: its pretty unnecessary to catch and raise this exception anyways in my opinion

File: torrent/tracker.py [hunk 28]
```diff
@@ -0,0 +1,378 @@
...
+    if wait > 1800:
+        wait = 1800
+    sleep(wait) 
+    while True:
+        logger.debug(f"start http tracker request")
+        res = None
+        try: 
+            res = requests.get(url)
+        except requests.exceptions.ConnectionError:
+            logger.debug("connect to tracker failed")
```
Comment (adarmaor) [inline torrent/tracker.py:284]: maybe you should work with a helper function that gets a function, exception and log message and does this whole thing for you

File: torrent/tracker.py [hunk 29]
```diff
@@ -0,0 +1,378 @@
...
+            tracker_threads.append(udp_request(tracker, torrent_file, peer_manager))
+        else:
+            #TODO add support to dht protocol
+            pass
+    return tracker_threads
+        
+
+
+
+if __name__ == "__main__":
```
Comment (adarmaor) [inline torrent/tracker.py]: remove this

File: torrent/tracker.py [hunk 30]
```diff
@@ -0,0 +1,378 @@
...
+
+def get_ip(ip_address):
+    return '.'.join([str(num) for num in ip_address])
+
+
+#request manager
+def tracker_manager(torrent_file, peer_manager):
+    tracker_threads = []
+    for tracker in torrent_file.trackers:
+        if tracker.tracker_type == 'http':
```
Comment (adarmaor) [inline torrent/tracker.py:351]: http and udp can be a common enum as well

File: torrent/utilities.py [hunk 31]
```diff
@@ -0,0 +1,53 @@
...
+    return wrap
+
+def od_get_key(od, key, mandat=False):
+    """
+    the function get OrederdDict object key and mandet
+    the function try to get back the data from the od 
+    if it mandat raise the error if the key is not mandat 
+    return None
+    """
+    if type(od) != OrderedDict:
```
Comment (adarmaor) [inline torrent/utilities.py]: why do you keep checking this?????

File: torrent/utilities.py [hunk 32]
```diff
@@ -0,0 +1,53 @@
...
+            pass
+        q = Queue() 
+        t = Thread(target=wrapped_f, args=(q, ) + args, kwargs=kwargs)
+        t.daemon = daemon
+        t.start()
+        t.result_queue = q
+        return t
+    return wrap
+
+def od_get_key(od, key, mandat=False):
```
Comment (adarmaor) [inline torrent/utilities.py]: change mandat to something more understood, what does it mean???

Comment (adarmaor) [reply torrent/utilities.py]: make it raise_on_failure or something

File: torrent/utilities.py [hunk 33]
```diff
@@ -0,0 +1,53 @@
...
+        raise TypeError(f"od_get_key get '{mandat}' for mandat variable the function get only bool type" )
+    try:
+        return od[key]
+    except KeyError as e:
+        if mandat:
+            logger.warning(f'od_get_key get({type(od)},{type(key)},{mandat}) and raise KeyError', exc_info=True)
+            raise e
+        else:
+            return None
+    except TypeError as e:
```
Comment (adarmaor) [inline torrent/utilities.py]: why would this raise a TypeError?

File: torrent/networking.py (hunk 13 above)
Comment (avrham408) [reply torrent/networking.py:88]: fixed

Comment (avrham408) [reply torrent/networking.py:88]: all this process rebuild in next verison

File: torrent/networking.py (hunk 14 above)
Comment (avrham408) [reply torrent/networking.py]: fixed

File: torrent/utilities.py (hunk 32 above)
Comment (avrham408) [reply torrent/utilities.py]: fixed my brother

File: tests/test_pieces.py [hunk 34]
```diff
@@ -34,6 +34,16 @@ def test_create_pieces_manager_last_piece():
...
+    for torrent_path in files:
+        torrent_file = generate_torrent_file(torrent_path)
+        piece_manager = create_piece_manager(torrent_file)
+        status_dict = piece_manager.pieces_status()
+        assert len(piece_manager.pieces) == sum([i for i in status_dict.values()])
+
+    
+
 
 if __name__ == "__main__":
```
Comment (adarmaor) [inline tests/test_pieces.py:48]: if you work with pytest you don't need this

File: torrent/torrent_file.py [hunk 35]
```diff
@@ -26,32 +26,28 @@ def __init__(self, struct_dict):
...
         self.raw_content = struct_dict['raw']
-        self.peer_id = '-PC0001-' + ''.join([str(randint(0, 9)) for _ in range(12)])
+        self.peer_id = f"-PC0001-{''.join([str(randint(0, 9)) for _ in range(12)])}"
 
     def __repr__(self):
         return f"torrent_file({self.name})"
 
 
 class Tracker():
-    #TODO think if this is the right place hold this object and if the the future the tracker going to
```
Comment (adarmaor) [inline torrent/torrent_file.py:26]: add this comment to the todo file

Comment (adarmaor) [reply torrent/torrent_file.py:26]: or remove it completely

**Pull request merged**